File: backend/pkg/pagination/pagination.go
package pagination

import (
	"math"
	"net/http"
	"strconv"
	"strings"
)

// Params holds parsed query parameters for pagination, searching, and sorting.
type Params struct {
	Page      int    `json:"page"`
	PerPage   int    `json:"per_page"`
	Search    string `json:"search"`
	Category  string `json:"category"`
	SortBy    string `json:"sort_by"`
	SortOrder string `json:"sort_order"`
}

// Offset returns the zero-indexed SQL offset.
func (p Params) Offset() int {
	return (p.Page - 1) * p.PerPage
}

// Meta holds pagination metadata to be returned in API responses.
type Meta struct {
	TotalItems  int64 `json:"total_items"`
	TotalPages  int   `json:"total_pages"`
	CurrentPage int   `json:"current_page"`
	PerPage     int   `json:"per_page"`
}

// Response represents a standard paginated API envelope.
type Response[T any] struct {
	Data []T  `json:"data"`
	Meta Meta `json:"meta"`
}

// Parse extracts and sanitizes pagination parameters from an HTTP request.
func Parse(r *http.Request) Params {
	q := r.URL.Query()

	page := 1
	if p, err := strconv.Atoi(q.Get("page")); err == nil && p > 0 {
		page = p
	}

	perPage := 10
	if l := q.Get("per_page"); l != "" {
		if val, err := strconv.Atoi(l); err == nil && val > 0 {
			perPage = val
		}
	} else if l := q.Get("limit"); l != "" {
		if val, err := strconv.Atoi(l); err == nil && val > 0 {
			perPage = val
		}
	}
	if perPage > 100 {
		perPage = 100
	}

	search := strings.TrimSpace(q.Get("search"))
	if search == "" {
		search = strings.TrimSpace(q.Get("q"))
	}

	sortOrder := strings.ToLower(strings.TrimSpace(q.Get("sort_order")))
	if sortOrder != "asc" && sortOrder != "desc" {
		sortOrder = "desc"
	}

	return Params{
		Page:      page,
		PerPage:   perPage,
		Search:    search,
		Category:  strings.TrimSpace(q.Get("category")),
		SortBy:    strings.TrimSpace(q.Get("sort_by")),
		SortOrder: sortOrder,
	}
}

// NewMeta calculates the total pages and builds a Meta object.
func NewMeta(totalItems int64, page, perPage int) Meta {
	totalPages := 0
	if totalItems > 0 && perPage > 0 {
		totalPages = int(math.Ceil(float64(totalItems) / float64(perPage)))
	}
	return Meta{
		TotalItems:  totalItems,
		TotalPages:  totalPages,
		CurrentPage: page,
		PerPage:     perPage,
	}
}

// PaginateSlice takes an in-memory slice and returns a paginated slice along with pagination metadata.
func PaginateSlice[T any](items []T, params Params) Response[T] {
	totalItems := int64(len(items))
	meta := NewMeta(totalItems, params.Page, params.PerPage)

	if totalItems == 0 {
		return Response[T]{
			Data: []T{},
			Meta: meta,
		}
	}

	start := params.Offset()
	if start >= len(items) {
		return Response[T]{
			Data: []T{},
			Meta: meta,
		}
	}

	end := start + params.PerPage
	if end > len(items) {
		end = len(items)
	}

	return Response[T]{
		Data: items[start:end],
		Meta: meta,
	}
}

File: frontend/src/lib/actions/pagination.ts
export interface PaginationParams {
  page?: number;
  per_page?: number;
  limit?: number;
  search?: string;
  category?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  mine?: boolean;
  [key: string]: any;
}

export interface PaginationMeta {
  total_items: number;
  total_pages: number;
  current_page: number;
  per_page: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  meta: PaginationMeta;
}

/**
 * Normalizes an API response that may be either a raw array or a paginated envelope { data, meta }.
 */
export function normalizePaginatedResponse<T>(
  res: any,
  fallbackPage = 1,
  fallbackPerPage = 10
): PaginatedResponse<T> {
  if (!res) {
    return {
      data: [],
      meta: {
        total_items: 0,
        total_pages: 0,
        current_page: fallbackPage,
        per_page: fallbackPerPage,
      },
    };
  }

  if (Array.isArray(res)) {
    return {
      data: res,
      meta: {
        total_items: res.length,
        total_pages: Math.ceil(res.length / fallbackPerPage) || (res.length > 0 ? 1 : 0),
        current_page: fallbackPage,
        per_page: fallbackPerPage,
      },
    };
  }

  const data = Array.isArray(res.data) ? res.data : [];
  const meta: PaginationMeta = {
    total_items: typeof res.meta?.total_items === "number" ? res.meta.total_items : data.length,
    total_pages: typeof res.meta?.total_pages === "number" ? res.meta.total_pages : (data.length > 0 ? 1 : 0),
    current_page: typeof res.meta?.current_page === "number" ? res.meta.current_page : fallbackPage,
    per_page: typeof res.meta?.per_page === "number" ? res.meta.per_page : fallbackPerPage,
  };

  return { data, meta };
}
